crates/ruff_linter/resources/test/fixtures/flake8_logging/LOG014.py

```diff
@@ -0,0 +1,154 @@
+logger.info("foo", exc_info=True)  # LOG014
+
+try:
+    a = 1 * 2
+except Exception:
+    logger.exception("foo")  # OK
+    logger.exception("foo", exc_info=False)  # OK
+    logger.exception("foo", exc_info=[])  # OK
+    logger.exception("foo", exc_info=True)  # OK
+    logger.exception("foo", exc_info=[1])  # OK
+
+    logger.error("foo", exc_info=False)  # OK
+    logger.error("foo", exc_info=True)  # OK
+
+    logger.info("foo", exc_info=False)  # OK
+    logger.info("foo", exc_info=True)  # OK
+
+# Direct Call ------------------------------------------------------------------
+
+from logging import exception, error, info
+
+exception("foo")  # OK
+exception("foo", exc_info=False)  # OK
+exception("foo", exc_info=[])  # OK
+exception("foo", exc_info=True)  # LOG014
+exception("foo", exc_info=[1])  # LOG014
+
+error("foo", exc_info=False)  # OK
+error("foo", exc_info=True)  # LOG014
+
+info("foo", exc_info=False)  # OK
+info("foo", exc_info=True)  # LOG014
+
+try:
+    a = 1 * 2
+except Exception:
+    exception("foo")  # OK
+    exception("foo", exc_info=False)  # OK
+    exception("foo", exc_info=[])  # OK
+    exception("foo", exc_info=True)  # OK
+    exception("foo", exc_info=[1])  # OK
+
+    error("foo", exc_info=False)  # OK
+    error("foo", exc_info=True)  # OK
+
+    info("foo", exc_info=False)  # OK
+    info("foo", exc_info=True)  # OK
+
+# Fake Call --------------------------------------------------------------------
+
+def wrapper():
+    exception = lambda *args, **kwargs: None
+    error = lambda *args, **kwargs: None
+    info = lambda *args, **kwargs: None
+
+    exception("foo")  # OK
+    exception("foo", exc_info=False)  # OK
+    exception("foo", exc_info=[])  # OK
+    exception("foo", exc_info=True)  # OK
+    exception("foo", exc_info=[1])  # OK
+
+    error("foo", exc_info=False)  # OK
+    error("foo", exc_info=True)  # OK
+
+    info("foo", exc_info=False)  # OK
+    info("foo", exc_info=True)  # OK
+
+    try:
+        a = 1 * 2
+    except Exception:
+        exception("foo")  # OK
+        exception("foo", exc_info=False)  # OK
+        exception("foo", exc_info=[])  # OK
+        exception("foo", exc_info=True)  # OK
+        exception("foo", exc_info=[1])  # OK
+
+        error("foo", exc_info=False)  # OK
+        error("foo", exc_info=True)  # OK
+
+        info("foo", exc_info=False)  # OK
+        info("foo", exc_info=True)  # OK
+
+# Nested -----------------------------------------------------------------------
+
+def apple_1():
+    try:
+        logging.exception("foo", exc_info=True)  # LOG014
+    except Exception:
+        pass
+
+
+def apple_2():
+    def banana():
+        logging.exception("foo", exc_info=True)  # LOG014
+
+    try:
+        banana()
+    except Exception:
+        pass
+
+
+def apple_3():
+    def banana():
+        logging.exception("foo", exc_info=True)  # LOG014
+
+    try:
+        a = 1 * 2
+    except Exception:
+        banana()
```

Comment:
The ecosystem checks are turning up a few instances just like this.

Technically in this case, if we are able to tell that `banana()` is *only* ever called from inside an exception handling block, then we could mark this as okay. But I don't think it's possible to check all calling sites?

Reply:
It's possible, but you'd need to move the rule to `crates/ruff_linter/src/checkers/ast/analyze/bindings.rs` rather than `crates/ruff_linter/src/checkers/ast/analyze/expression.rs`. The difference is that the rules executed from `bindings.rs` are run after the entire semantic model has been built, so they have more information available to them in some ways: they're able to iterate through all the references to a binding as well as looking at the original binding itself. You can see an example of how to do it in a PR I merged this morning: https://github.com/astral-sh/ruff/pull/14661.

Reply:
Ahh. I'll revise it to include that check then. Thanks.

Reply:
no problem, thanks for the PR!

Reply:
Forgive me if I'm misunderstanding, but I don't think moving it to `bindings.rs` works?

An expression like `logging.info("foo", exc_info=True)` isn't a binding, so if I move `exc_info_outside_exception_handler()` to `bindings.rs`, then the only situation the lint could possibly apply would be `x = logging.info("foo", exc_info=True)` or `(x := logging.info("foo", exc_info=True))`, which no one writes.

Is it possible to make a new version of `expression.rs` that's called after the entire semantic model has been built? I don't think there's any way around calling the lint function on expressions.

Reply:
> An expression like `logging.info("foo", exc_info=True)` isn't a binding, so if I move my lint function call to `bindings.rs`, then the only situation the lint could possibly apply would be `x = logging.info("foo", exc_info=True)` or `(x := logging.info("foo", exc_info=True))`, which no one writes.

ah, you're quite right.

I suppose what would be best would be to collect all `logging.(...)` calls into a `Vec` stored on [this struct](https://github.com/astral-sh/ruff/blob/6dfe125f44352f5fdff19f6d47fc33ac7a6e86ad/crates/ruff_linter/src/checkers/ast/deferred.rs#L4-L15) as we traverse the AST in https://github.com/astral-sh/ruff/blob/main/crates/ruff_linter/src/checkers/ast/mod.rs. Then after the AST has been fully visited and the `SemanticModel` has been fully constructed, we can call this rule on each `logging(...)` call we stored in the Vec.